Roadmap v2 (#410)
paritytech/zombienet: roadmap.md

Roadmap Zombienet v2

> ## Infra

Reviewer: Prometheus server deployment -> matching against prometheus queries. 
- we remove code and inefficiency from scraping metrics ourselves
- scales better
- we get a standardized way of querying metrics and ability to do aggregations as we see fit
- (nice to have) create some test reports using these, maybe alarms for long duration failure conditions
- local debugability increases

Open question:
- do we use prometheus for native/local runs, or just k8s ? (if this differers across providers, prometheus queries won't work)

Author: I think local prometheus should be `opt-in` and not a default.

> ## Infra

Reviewer: ### Long lived test networks deployment and management.
- deploy and manage Versi using Zombienet - similar functionality with validator manager
- obsoletes our custom solution currently built and maintained by devops

> ## Infra

Reviewer: Deployment scalability improvements - up to 1000 validators.

> - Chaos testing, add examples and explore possibilities in `native` and `podman` provider

Reviewer: This is top priority for parachains. We want to roll out a separate CI pipeline to do these long duration tests.

> - Add docker provider
> - Add nomad provider

Reviewer 2: ```suggestion
- Add `docker` provider
- Add `nomad` provider
```

> - Add `nomad` provider

Reviewer 3: :+1:

Author: Hi @drahnr, did you have time to make a quick chat to review some request relates to nomad?
Thanks!

> - Create helm chart to allow other use zombienet in k8s

Reviewer 2: ```suggestion
- Create [helm chart](https://helm.sh/docs/topics/charts/) to allow other use zombienet in k8s
```

- Auth system to not use k8s users

> - Create GitHub Action and publish in marketplace

Reviewer 2: Are you referring to npmjs here?
If yes then you can commit the following suggestion:

```suggestion
- Create GitHub Action and publish in marketplace (wip)
```

> ## Intenal teams

Reviewer: ```suggestion
## Internal teams
```

- Add more teams (wip)

## UI

> - Create UI to create `feature` and `network` file.

Reviewer 2: ```suggestion
- Create UI to create `.feature` (`.zndls`) and `.network` files.
```

- Create script to translate from polkadot launch config to zombienet network file.

> - Improve vscode extension (grammap/snippets)

Reviewer 2: ```suggestion
- Improve vscode extension (grammar/snippets)
```

> - Create script to translate from polkadot launch config to zombienet network file.
> - Improve vscode extension (grammar/snippets)

Reviewer 2: ```suggestion
- Create script to translate from polkadot launch config to zombienet network file. ([wip](https://github.com/paritytech/zombienet/tree/nik-convert-polkadot-launch-config-to-zombienet))
- Improve VSCode extension (grammar/snippets/syntax highlighting) ([repo](https://github.com/paritytech/zombienet-vscode-extension))
```

## Registry
- Create decorators registry and allow override by paras (wip)
- Explore how to get info from paras.

> ## Functional tasks

Reviewer: Zombienet Test SDK (Rust) - it's not replacing the deployment which remains as is (typescript) + additional features listed above.
- one language to fully write the entire test (deployment part + actual tests to be performed)
- simple APIs to generate the `Network` file - builder pattern
- simple APIs to query/assert metrics
- write more complex test logic
- directly use or wrap subxt - open HRMP channels, send XCM, do runtime upgrades, etc
- opens up the Rust ecosystem of libraries that can be used in a test 
- feature parity with current DSL support (phasing out DSL over time after we have same featureset)
- easier for people outside Parity to contribute to the project

> 

Reviewer 2: - Add UI app (desktop) to run zombienet without the need of terminal (??)

- Add subxt integration, allow to compile/run on the fly
- Move parser to pest (wip)

> - Dettach phases and use json to communicate instead of `paths`

Reviewer 2: ```suggestion
- Detach phases and use JSON to communicate instead of `paths`
```

> - Add UI app (desktop) to run zombienet without the need of terminal 

Reviewer 2: ```suggestion
- Create UI app (desktop) to run zombienet without the need of terminal.

```

> - Move parser to pest (wip)

Reviewer 3: :heart: happy to collab on this :)

Author: Hey @drahnr, I started the work in this [draft pr](https://github.com/paritytech/zombienet/pull/382). The parser is already working and I need to finish the wiring between the parser crate and the test-runner. If you have time to review the pr would be awesome :)

- Create script to translate from polkadot launch config to zombienet network file. ([wip](https://github.com/paritytech/zombienet/tree/nik-convert-polkadot-launch-config-to-zombienet))
- Improve VSCode extension (grammar/snippets/syntax highlighting) ([repo](https://github.com/paritytech/zombienet-vscode-extension))
- Create UI app (desktop) to run zombienet without the need of terminal.

## Registry
- Create decorators registry and allow override by paras (wip)
- Explore how to get info from paras.

## Functional tasks
- Add subxt integration, allow to compile/run on the fly
- Move parser to pest (wip)
- Detach phases and use JSON to communicate instead of `paths`
- Add relative values assertions (for metrics/scripts)
- Allow to define nodes that are not started in the launching phase and can be started by the test-runner
- Allow to define `race` assertions
- Rust integration -> Create multiples libs (crates)
- Explore backchannel use case
- Add support to run test agains a running network (wip)
- Add more CLI subcommands
- Add js/subxt snippets ready to use in assertions (e.g transfers)
- Add XCM support in built-in assertions

> - Add support to start from a live network (fork-off) [check subalfred]

Reviewer 2: ```suggestion
- Add support to start from a live network (fork-off) [check subalfred]
- Create "default configuration" - (if `zombieconfig.json` exists in same dir with zombienet then the config applied in it will override the default configuration of zombienet.  E.G if user wants to have as default `native` instead of `k8s` he can add  to default the entry `{ default: 'native' }`.)
```

Roadmap Zombienet v2

## Infra
- Chaos testing, add examples and explore possibilities in `native` and `podman` provider
- Add `docker` provider
- Add `nomad` provider
- Create [helm chart](https://helm.sh/docs/topics/charts/) to allow other use zombienet in k8s
- Auth system to not use k8s users

> - Create GitHub Action and publish in marketplace (wip)

Reviewer 2: ```suggestion
- Create GitHub Action and publish in marketplace (wip)
- Rename `@paritytech/zombienet` npm package to `zombienet`. Keep all zombienet modules under `@zombienet/*` org
```

> - Create GitHub Action and publish in marketplace (wip)

Reviewer 2: ```suggestion
- Create GitHub Action and publish in NPM marketplace (Completed)
```

> - Rename `@paritytech/zombienet` npm package to `zombienet`. Keep all zombienet modules under `@zombienet/*` org

Reviewer 2: ```suggestion
- Rename `@paritytech/zombienet` npm package to `zombienet` (wip)
- Keep all zombienet modules under `@zombienet/*` org (Completed)
```

> 

Reviewer 2: ```suggestion
- Add functionality of running different images of same relayChain/parachain (can be helpful for backward compatibility - [related issue](https://github.com/paritytech/zombienet/issues/457]))

```

## Internal teams
- Add more teams (wip)

## UI

> - Create UI to create `.feature` (`.zndls`) and `network` files.

Reviewer 2: ```suggestion
- Create UI to create `.zndls` and `network` files.
```

> - Create script to translate from polkadot launch config to zombienet network file. ([wip](https://github.com/paritytech/zombienet/tree/nik-convert-polkadot-launch-config-to-zombienet))

Reviewer 2: ```suggestion
- Create script to translate from polkadot launch config to zombienet network file. [Open PR](https://github.com/paritytech/zombienet/tree/nik-convert-polkadot-launch-config-to-zombienet) (once merged we can remove from roadmap)
```

> - Create script to translate from polkadot launch config to zombienet network file. [Open PR](https://github.com/paritytech/zombienet/tree/nik-convert-polkadot-launch-config-to-zombienet) (once merged we can remove from roadmap)

Reviewer 2: This is done and PR is merged. Should we remove it from the roadmap already @pepoviola ?
```suggestion
```

Author: Yes, thanks @wirednkod

> - Improve VSCode extension (grammar/snippets/syntax highlighting) ([repo](https://github.com/paritytech/zombienet-vscode-extension))

Reviewer 2: ```suggestion
- Improve VSCode extension (grammar/snippets/syntax highlighting/file validations) ([repo](https://github.com/paritytech/zombienet-vscode-extension))
```

- Create UI app (desktop) to run zombienet without the need of terminal.

## Registry
- Create decorators registry and allow override by paras (wip)
- Explore how to get info from paras.

## Functional tasks
- Add subxt integration, allow to compile/run on the fly
- Move parser to pest (wip)
- Detach phases and use JSON to communicate instead of `paths`
- Add relative values assertions (for metrics/scripts)
- Allow to define nodes that are not started in the launching phase and can be started by the test-runner
- Allow to define `race` assertions
- Rust integration -> Create multiples libs (crates)
- Explore backchannel use case
- Add support to run test agains a running network (wip)
- Add more CLI subcommands
- Add js/subxt snippets ready to use in assertions (e.g transfers)

> - Add XCM support in built-in assertions

Reviewer 2: Add `ink! smart contract` support

cc @pepoviola

> - Add support to start from a live network (fork-off) [check subalfred]

Reviewer 4: From our experience (already doing fork-off Moonbeam chains):
This will require chain-specific actions (changing the authoring values, changing council, changing balance....). So a proper way to interact with the raw state is needed for each case.
It also requires to deal with very large database (or exported state json file) (Moonbeam > 5GB), which is not often easily done.

Reviewer 5: We should perhaps file sub-issues to all roadmap items for further discussion once finalized. 

Certainly, this is a desirable feature. The main motivation I see is to test node or runtime changes against live network states as a pre-release check. This does require some ability to interact with raw state. To start from a live network state, the two major things are to allow swapping out the runtime (optional) and to allow modification of the state.

Author: > We should perhaps file sub-issues to all roadmap items for further discussion once finalized.
> 
> Certainly, this is a desirable feature. The main motivation I see is to test node or runtime changes against live network states as a pre-release check. This does require some ability to interact with raw state. To start from a live network state, the two major things are to allow swapping out the runtime (optional) and to allow modification of the state.

Thanks for your feedback @rphmeier! I will create a couple of umbrella issues to capture the vision/plan for the public roadmap and individual issues for items.

Thanks!
